clang/lib/Sema/SemaBounds.cpp

```diff
@@ -3811,6 +3860,142 @@ namespace {
+        << DeclaredBounds << DeclaredBounds->getSourceRange();
+
+      // The observed bounds of v are unknown because the original observed
+      // bounds B of v used a variable w, and there was an assignment to w
+      // where w had no original value.
+      auto LostVarIt = State.LostVariables.find(V);
+      if (LostVarIt != State.LostVariables.end()) {
+        std::pair<BoundsExpr *, DeclRefExpr *> Lost = LostVarIt->second;
+        BoundsExpr *InitialObservedBounds = Lost.first;
+        DeclRefExpr *LostVar = Lost.second;
+        S.Diag(LostVar->getLocation(), diag::note_lost_variable)
+          << LostVar << InitialObservedBounds << V << LostVar->getSourceRange();
+      }
+
+      // The observed bounds of v are unknown because at least one expression
+      // e with unknown bounds was assigned to v.
+      auto BlameSrcIt = State.UnknownSrcBounds.find(V);
+      if (BlameSrcIt != State.UnknownSrcBounds.end()) {
+        SmallVector<Expr *, 4> UnknownSources = BlameSrcIt->second;
+        for (auto I = UnknownSources.begin(); I != UnknownSources.end(); ++I) {
+          Expr *Src = *I;
+          S.Diag(Src->getBeginLoc(), diag::note_unknown_source_bounds)
+            << Src << V << Src->getSourceRange();
+        }
+      }
+    }
+
+    // CheckObservedBounds checks that the observed bounds for a variable v
+    // imply that the declared bounds for v are provably true after checking
+    // the top-level CFG statement St.
+    //
+    // EquivExprs contains all equality facts contained in State.EquivExprs,
+    // as well as any equality facts implied by State.TargetSrcEquality.
+    void CheckObservedBounds(Stmt *St, const VarDecl *V,
+                             BoundsExpr *DeclaredBounds,
+                             BoundsExpr *ObservedBounds, CheckingState State,
+                             EquivExprSets *EquivExprs,
+                             CheckedScopeSpecifier CSS) {
+      ProofFailure Cause;
+      ProofResult Result = ProveBoundsDeclValidity(DeclaredBounds, ObservedBounds,
+                                                   Cause, EquivExprs);
+      if (Result == ProofResult::True)
+        return;
+
+      // If v currently has widened bounds, then the proof failure was caused
+      // by not being able to prove the widened bounds of v imply the declared
+      // bounds of v.  Diagnostics should not be emitted in this case.
+      // Otherwise, statements that make no changes to v or any variables used
+      // in the bounds of v would cause diagnostics to be emitted.
+      // For example, the widened bounds (p, (p + 0) + 1) do not provably imply
+      // the declared bounds (p, p + 0) due to the left-associativity of the
+      // observed upper bound (p + 0) + 1.
+      // TODO: checkedc-clang issue #867: the widened bounds of a variable
+      // should provably imply the declared bounds of a variable.
+      if (State.WidenedVariables.find(V) != State.WidenedVariables.end())
```

Comment:
Widened bounds currently do not always provably imply declared bounds. Opened issue #867 to track this.